pkg/volume/iscsi/iscsi_util.go

```diff
@@ -695,31 +733,33 @@ func (util *ISCSIUtil) DetachBlockISCSIDisk(c iscsiDiskUnmapper, mapPath string)
 		if err != nil {
 			return err
 		}
-		arr := strings.Split(device, "-lun-")
-		if len(arr) < 2 {
-			return fmt.Errorf("failed to retrieve lun from mapPath: %v", mapPath)
-		}
-		lun = arr[1]
 		// Extract the iface from the mountPath and use it to log out. If the iface
 		// is not found, maintain the previous behavior to facilitate kubelet upgrade.
 		// Logout may fail as no session may exist for the portal/IQN on the specified interface.
 		iface, found = extractIface(mapPath)
 	}
+
+	// Delete all the scsi devices and any multipath devices
+	if err = deleteDevices(c.iscsiDisk, c.exec, c.deviceUtil); err != nil {
+		klog.Warningf("iscsi detach disk: failed to delete devices\nError: %v", err)
+		// Fall through -- even if deleting fails, a logout may fix problems
+	}
+
+	// Lock the target while we determine if we can safely log out or not
+	c.plugin.targetLocks.LockKey(iqn)
+	defer c.plugin.targetLocks.UnlockKey(iqn)
+
+	// if device is no longer used, see if need to logout the target
+	refCount, err := getDevicePrefixRefCount(c.mounter, mapPath, ioutil.ReadDir, prefix)
+	if err != nil || refCount != 0 {
+		return nil
+	}
+
 	portals := removeDuplicate(bkpPortal)
 	if len(portals) == 0 {
 		return fmt.Errorf("iscsi detach disk: failed to detach iscsi disk. Couldn't get connected portals from configurations")
 	}
 
-	devicePath := getDevByPath(portals[0], iqn, lun)
-	klog.V(5).Infof("iscsi: devicePath: %s", devicePath)
-	if _, err = os.Stat(devicePath); err != nil {
```

Comment:
the block detach path tries to "validate devicePath" but the filesystem does not. Why might it be needed? I don't think it is, so I am removing it.